gobblin-runtime/src/main/java/gobblin/runtime/util/ParallelStateSerDeRunner.java

```diff
@@ -0,0 +1,231 @@
+   * </p>
+   *
+   * @param state an empty {@link State} object to which the deserialized content will be populated
+   * @param inputFilePath the input file to read from
+   * @param <T> the {@link State} object type
+   * @throws IOException if there's anything deserializing the {@link State} object
+   */
+  public <T extends State> void deserializeFromFile(final T state, final Path inputFilePath) throws IOException {
+    this.futures.add(this.executor.submit(new Callable<Void>() {
+
+      @Override
+      public Void call()
+          throws Exception {
+        Closer closer = Closer.create();
+        try {
+          InputStream inputStream = closer.register(fs.open(inputFilePath));
+          DataInputStream dataInputStream = closer.register(new DataInputStream(inputStream));
+          state.readFields(dataInputStream);
+        } catch (Throwable t) {
+          throw closer.rethrow(t);
+        } finally {
+          closer.close();
+        }
+
+        return null;
+      }
+    }));
+  }
+
+  /**
+   * Deserialize a list of {@link State} objects from a Hadoop {@link SequenceFile}.
+   *
+   * <p>
+   *   This method submits a task to deserialize the {@link State} objects and returns immediately
+   *   after the task is submitted.
+   * </p>
+   *
+   * @param stateClass the {@link Class} object of the {@link State} class
+   * @param inputFilePath the input {@link SequenceFile} to read from
+   * @param states a {@link Collection} object to store the deserialized {@link State} objects
+   * @param <T> the {@link State} object type
+   * @throws IOException if there's anything deserializing the {@link State} objects
+   */
+  public <T extends State> void deserializeFromSequenceFile(final Class<? extends Writable> keyClass,
+      final Class<T> stateClass, final Path inputFilePath, final Collection<T> states) throws IOException {
+    this.futures.add(this.executor.submit(new Callable<Void>() {
+      @Override
+      public Void call()
+          throws Exception {
+        Closer closer = Closer.create();
+        try {
+          SequenceFile.Reader reader = closer.register(new SequenceFile.Reader(fs, inputFilePath, fs.getConf()));
+          Writable key = keyClass.newInstance();
+          T state = stateClass.newInstance();
+          while (reader.next(key, state)) {
+            states.add(state);
+            state = stateClass.newInstance();
+          }
+        } catch (Throwable t) {
+          throw closer.rethrow(t);
+        } finally {
+          closer.close();
+        }
+
+        return null;
+      }
+    }));
+  }
+
+  /**
+   * Wait until the {@link ParallelStateSerDeRunner} is done running all submitted tasks.
+   *
+   * @throws InterruptedException if the calling thread is interrupted while waiting
+   * @throws ExecutionException if any of the task throws an exception during execution
+   */
+  public void awaitDone() throws InterruptedException, ExecutionException {
```

Comment:
Can't we have the awaitDone() logic as part of close() ? This way we don't have to worry if awaitDone() has been called before close().